fix: enhance FocusDirective with cxRefreshFocusOn input

wangzixi-diablo opened the pull request:
closes: #9531

wangzixi-diablo committed:
enhance FocusDirective with cxRefreshFocusOn input

wangzixi-diablo committed:
fix build error caused by wrong file path

wangzixi-diablo commented:
As a draft PR, currently only focusable element in Unit detail page (unit-details.component.html) is automatically focused. Wait for expert's feedback and then continue.

wangzixi-diablo commented:
As a draft PR, currently only focusable element in Unit detail page (unit-details.component.html) is automatically focused. Wait for expert's feedback and then continue.

# how to test

Choose one table row in Unit list page:

![image](https://user-images.githubusercontent.com/58975336/102779433-2c922a80-43cf-11eb-9997-e2184e6d04ae.png)

![image](https://user-images.githubusercontent.com/58975336/102779513-54818e00-43cf-11eb-8919-75e426852670.png)

When the detail page is already rendered, change the selection of unit, and detail page will be refreshed with new unit data loaded from backend. After detail page is refreshed, again Edit element will be focused automatically:

![image](https://user-images.githubusercontent.com/58975336/102779657-93174880-43cf-11eb-9306-b3e2888afc8d.png)

# drawback of this enhancement

Take this example for illustration: when the detail page is rendered for the first time, Edit element will be focused by code **TWICE**:

**Firstly**, it's focused because of the following code in focus.directive.ts:

```typescript
@Input() set cxRefreshFocusOn(_switchCondition: string) {
    if (this.shouldAutofocus) {
      super.handleFocus(); -------------------- this line focuses the Edit element
    }
  }
```

**Secondly**, it's focused because of the following code in auto-focus.directive.ts:

```typescript
ngAfterViewInit(): void {
    if (this.shouldAutofocus) {
      this.handleFocus(); -------------------- this line focuses the Edit element
    }
    if (!this.shouldAutofocus || this.hasPersistedFocus) {
      super.ngAfterViewInit();
    }
  }
```

After unit detail view is rendered, its ngAfterViewInit will not be triggered, thus no duplicated focus will occur.

tobi-or-not-tobi commented on feature-libs/organization/administration/components/unit/details/unit-details.component.html:
```diff
@@ -1,12 +1,19 @@
 <cx-org-card *ngIf="model$ | async as model" i18nRoot="orgUnit.details">
-  <a
+  <div
```
I'd not introduce a new DOM element here, you can apply the condition and config directly to the anchor.

wangzixi-diablo replied:
Hi Tobias, 

Thanks a lot for your kind review! Let me explain why I introduce a wrapper div element on purpose.

Yes, I also observed that the host element which declares [cxFocus] directive will have tabindex set as -1. 

I debugged and knew the reason, tabindex is set as -1 by default in base-focus.directive.ts:

![image](https://user-images.githubusercontent.com/58975336/103994974-bb29e880-51d2-11eb-8330-0c21290cd5cb.png)

In order to avoid this issue, I introduce a wrapper div element, and declare [cxFocus] directive on it. In this way, div itself has tabindex = -1, while edit button, the a tag, is still focusable:

![image](https://user-images.githubusercontent.com/58975336/103995256-3095b900-51d3-11eb-9b14-6ffb86496df6.png)

So do you know if there's any approach, to avoid the "side-effect" of having host element's tabindex attribute set as default value -1? I do believe there's a way, otherwise it does not make sense. 

![lock+Focus (1)](https://user-images.githubusercontent.com/58975336/103995488-85393400-51d3-11eb-9edd-038d29bf1fcc.png)

I tried to find some hint in lock-focus.directive.ts but unfortunately didn't find what I want - in my eye its logic is complex, I draw a mindmap to try to understand it...

Thanks a lot for your help!

tobi-or-not-tobi replied:
understood. I'd still like to avoid the div, can you debut on why `requiresExplicitTabIndex` is returning `true` in this case?

wangzixi-diablo replied:
Hi Tobias, 

I am fine with removal of this unnecessary div wrapper. Now <a> tag looks like below. I didn't add refresh function since it is not related to "tabindex = -1" trouble shooting.

```html
<a
      actions *ngIf="model.active"
      class="link edit"
      [class.disabled]="!model.active || (isInEditMode$ | async)"
      [routerLink]="{ cxRoute: 'orgUnitEdit', params: model } | cxUrl"
      [cxFocus]="{ autofocus: true }"
    >
      {{ 'organization.edit' | cxTranslate }}
    </a>
```

when debugging into function **requiresExplicitTabIndex**, it returns true because evaluation '**(this.host.hasAttribute('href') || this.host.hasAttribute('routerlink'))**' returns false.

![image](https://user-images.githubusercontent.com/58975336/104019239-289b4080-51f6-11eb-88e1-7bdd11aff871.png)

In the context of **requiresExplicitTabIndex** execution, <a> tag does not have href attribute populated yet. 

Neither does it has "routerlink", since at this time, "routerlink" is already converted to **ng-reflect-router-link** by Angular.

![image](https://user-images.githubusercontent.com/58975336/104019460-8def3180-51f6-11eb-9edf-3a8310ec09cb.png)

And as far as I know, href attribute of <a> tag only gets generated when another round of Angular change detection is performed. Unfortunately at this time the evaluation of [cxFocus] is already done, no chance to re-evaluated...

![image](https://user-images.githubusercontent.com/58975336/104019593-c858ce80-51f6-11eb-9e6e-7cc55e729b12.png)

# My Summary

I would describe this behavior as that [cxFocus] has a time dependency with [routerlink], what do you think? 

Of course, if I replace 'routerlink' with 'ng-reflect-router-link' in below code, the whole scenario will work, no tabindex=-1 any more. However since ng-reflect-router-link is internally populated by Angular, I don't think it's a practice to introduce it in our own code. 

![image](https://user-images.githubusercontent.com/58975336/104020284-e8d55880-51f7-11eb-997f-5db4527625ff.png)

tobi-or-not-tobi replied:
Good summary, and i'm not against using that angular internal to be frank. We could do a better job, but it takes a refactor that we cannot easily do without breaking changes. I've used the ng internal for now to unblock us with this behaviour. It's hitting us in more locations, so i'm eager to get this fixed.

tobi-or-not-tobi commented on feature-libs/organization/administration/components/unit/details/unit-details.component.html:
```diff
@@ -1,12 +1,19 @@
-    [routerLink]="{ cxRoute: 'orgUnitEdit', params: model } | cxUrl"
+    *ngIf="model.active"
+    [cxFocus]="{ autofocus: true }"
+    [cxRefreshFocusOn]="model.uid"
```
The cxFocus comes with a configuration. I'd consider extending the `AutoFocusConfig` with a `refreshFocus` property.

tobi-or-not-tobi replied:
worth trying if that gives the right result though.

tobi-or-not-tobi reviewed (commented):
This is a great start. During testing, I noticed that the edit button gets a -1 tabindex. This means that after leaving the element and tabbing back, you can't select it anymore. Something is off.

wangzixi-diablo committed:
remove wrapper div and avoid tabindex=-1 issue

wangzixi-diablo committed:
fix unit test error

wangzixi-diablo commented:
1. move this.requiredtabindex = -1 into ngAfterViewInit, due to two reasons:

a. before change: In the context of requiresExplicitTabIndex execution, tag does not have href attribute populated yet. For detail see this discussion [thread](https://github.com/SAP/spartacus/pull/10507#discussion_r553936239)

as a result, we have to put it to ngOnAfterViewInit, at that time the href attribute of a tag is already populated with correct value：

<img width="908" alt="clipboard1" src="https://user-images.githubusercontent.com/58975336/104442502-1cd1c480-55d0-11eb-8cde-4dd3f160486f.png">


b. The problem here is that we have a situation where the view generation process (which ngAfterViewInit is a part of) is itself further modifying the data that we are trying to display in the first place.

In unit test code it reports error message:

> ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'undefined'. Current value: '-1'.

See this [blog](https://blog.angular-university.io/angular-debugging/).

So I use setTimeout() to defer the code to another Javascript Virtual Machine turn.

2. Since now this.requiredtabindex = -1 is wrapped to setTimeout and executed asynchronously, some unit test will fail:

<img width="895" alt="clipboard2" src="https://user-images.githubusercontent.com/58975336/104442520-23603c00-55d0-11eb-9ed9-7866d91d3423.png">



3. how to use this extended attribute in units-detail.component.html?

<img width="885" alt="clipboard3" src="https://user-images.githubusercontent.com/58975336/104442545-2c510d80-55d0-11eb-96e4-52a9ab6dc874.png">

tobi-or-not-tobi commented on projects/storefrontlib/src/layout/a11y/keyboard-focus/base/base-focus.directive.ts:
```diff
@@ -38,10 +39,14 @@ export abstract class BaseFocusDirective implements OnInit {
+    /* to avoid unit test error:
+    Failed: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked
+    */
+    setTimeout(() => (this.requiredTabindex = -1), 0);
```
we don't like this in our code. There are other tricks to make unit tests work. You might want to override the change detection strategy in unit tests.

tobi-or-not-tobi commented on projects/storefrontlib/src/layout/a11y/keyboard-focus/focus.directive.ts:
```diff
@@ -11,6 +11,12 @@ export class FocusDirective extends LockFocusDirective {
   // tslint:disable-next-line: no-input-rename
   @Input('cxFocus') protected config: FocusConfig = {};
 
+  @Input() set cxRefreshFocusOn(_switchCondition: string) {
```
I'd not introduce this new input, but rather introduce another property in the `FocusConfig`

wangzixi-diablo committed:
use ngOnChanges hook to auto focus

wangzixi-diablo committed:
add unit test for refresh focus function

wangzixi-diablo commented:
# Some background of my latest commit ( 2021-2-2)

1. handleFocus method is called twice in different hooks in different focus class, which have different focus:

* in focus.directive.ts, ngOnChanges:

this hook is responsible to refresh focus whenever component property configured in focus configuration ( config.refreshFocus ) is changed.

* in auto-focus.directive.ts, ngOnAfterViewInit:

this hook is called when the view is fully rendered, responsible to auto-focus an element with autofocus = true.

In real scenario, for example when we opened unit list in B2B, click one unit to reach unit detail page(unit-details.component.html), an element marked with [cxFocus]="{ autofocus: true, refreshFocus: model.uid }" will be focused TWICE, one in ngOnChanges and the other one in ngAfterViewInit hook.

2. I didn't implement OnChanges in base-focus.directive.ts, but in focus.directive.ts instead.

The reason is, I checked in base-focus.directive.ts, the ngOnInit hook implemented there has some default logic. So if I implement OnChanges hook in base-focus.directive.ts, I don't have any default logic to code there, thus will lead to warnings in Visual Studio Code. So I choose to implement OnChanges in focus.directive.ts instead, this is current design in my PR.

wangzixi-diablo committed:
Merge branch 'develop' into feature/GH-9531

wangzixi-diablo committed:
fix broken unit test after resolve conflict with develop branch

wangzixi-diablo committed:
fix validation errors in travis CI build

tobi-or-not-tobi committed:
Improve refreshFocus

- dx: docs
- allow for autofocus requirement in case of refreshFocus
- recognise dynamic router based links during OnInit
- move `ngOnChanges` to proper location

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9531

Pucek9 committed:
Merge branch 'develop' of https://github.com/SAP/cloud-commerce-spartacus-storefront into feature/GH-9531

Pucek9 committed:
revert unexpected changes (wrong solved conflicts)

Pucek9 commented on feature-libs/organization/administration/components/unit/details/unit-details.component.html:
```diff
@@ -1,9 +1,11 @@
   [showHint]="true"
 >
   <a
+    *ngIf="model.active"
```
Is this change necessary? We are using already class=disabled for that unable to open edit form.
I tnink it can be breaking change and doesnt look good for mobile devices
```suggestion
```

tobi-or-not-tobi replied:
probably not actually. This was apparently added previously to mitigate a behaviour that I fixed in the focus directive. I wasn't aware of the ngif being added.

Pucek9 committed:
revert unexpected changes (wrong solved conflicts) part 2

Pucek9 committed:
apply changes to other details pages
changed placement of directive from org-xxx-form to form
fix focus for create
added focus for sublist

Pucek9 committed:
Update feature-libs/organization/administration/components/unit/details/unit-details.component.html

remove unnecessary change

Pucek9 committed:
Merge branch 'develop' into feature/GH-9531
